engine/src/bmr/BufferProvider.cpp

```diff
@@ -156,7 +158,22 @@ void allocation_pool::grow() {
 
 void allocation_pool::free_chunk(std::unique_ptr<blazing_allocation_chunk> buffer) {
   std::unique_lock<std::mutex> lock(in_use_mutex);
+  const std::size_t idx = buffer->allocation->index;
   buffer->allocation->allocation_chunks.push(std::move(buffer));
+
+  // if when we add the chunk back to the allocation, it now has all the allocations it needs, they free the allocation
+  // BUT only free the allocation if its not the very first allocation
+  // freeing the allocation means freeing the memory and deleting it from the allocation pool
+  // make sure things are "locked" when doing that to make sure its thread safe
+  if (idx > 0) {
+    if (this->allocations.at(idx)->total_number_of_chunks == this->allocations.at(idx)->allocation_chunks.size()) {
+        auto it = this->allocations.begin();
```

Comment:
`this->allocations.begin()` is pointing at the first element. The element you would want to delete is `idx`.

Reply:
the next line is an std::advance(it, idx) ... so this way I make the iterator point to that index